File: products/amliq/api/internal/ai/guard.go
package ai

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"time"
)

// guardResult represents the classification from the Claw guard endpoint.
type guardResult struct {
	Classification string   `json:"classification"`
	ViolationTypes []string `json:"violationTypes"`
}

var guardHTTP = &http.Client{Timeout: 2 * time.Second}

// guardInput checks input against the self-hosted Claw /v1/guard endpoint.
// Returns error if blocked, nil if passed or guard is unconfigured (fail-open).
func guardInput(ctx context.Context, input string) error {
	url := os.Getenv("CLAW_GATEWAY_URL")
	key := os.Getenv("CLAW_API_KEY")
	if url == "" || key == "" || input == "" {
		return nil
	}

	body, _ := json.Marshal(map[string]string{"input": input})
	req, err := http.NewRequestWithContext(ctx, "POST", url+"/v1/guard", bytes.NewReader(body))
	if err != nil {
		return nil
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "Bearer "+key)

	resp, err := guardHTTP.Do(req)
	if err != nil {
		return nil
	}
	defer resp.Body.Close()

	if resp.StatusCode != 200 {
		return nil
	}

	data, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil
	}

	var result guardResult
	if err := json.Unmarshal(data, &result); err != nil {
		return nil
	}

	if result.Classification == "block" {
		return fmt.Errorf("guard blocked: %v", result.ViolationTypes)
	}
	return nil
}

File: products/opensyber/apps/tokenforge-api/src/middleware/guard.ts
/**
 * Guard middleware — validates request bodies against prompt injection
 * via the self-hosted Claw gateway /v1/guard endpoint.
 *
 * Scans string fields in POST/PUT/PATCH bodies. Blocks requests where
 * any field is classified as a prompt injection attempt.
 *
 * Fail-open: if the guard endpoint is unreachable, requests proceed.
 */

import type { Context, Next } from 'hono';
import type { Env, Variables } from '../types.js';

const GUARD_TIMEOUT_MS = 2000;

/** Extract all string values from a request body for scanning */
function extractStrings(body: Record<string, unknown>): string[] {
  const strings: string[] = [];
  for (const value of Object.values(body)) {
    if (typeof value === 'string' && value.length > 3) {
      strings.push(value);
    }
  }
  return strings;
}

/**
 * Hono middleware — scans POST/PUT/PATCH bodies through Claw guard.
 * Attach after tenantAuth on /v1/* routes.
 */
export function guardMiddleware() {
  return async (c: Context<{ Bindings: Env; Variables: Variables }>, next: Next) => {
    const method = c.req.method;
    if (method !== 'POST' && method !== 'PUT' && method !== 'PATCH') {
      return next();
    }

    const guardUrl = (c.env as unknown as Record<string, string>).CLAW_GATEWAY_URL;
    const guardKey = (c.env as unknown as Record<string, string>).CLAW_API_KEY;
    if (!guardUrl || !guardKey) return next();

    let body: Record<string, unknown>;
    try {
      body = await c.req.json();
    } catch {
      return next();
    }

    const inputs = extractStrings(body);
    if (inputs.length === 0) return next();

    const combined = inputs.join('\n---\n');

    try {
      const controller = new AbortController();
      const timeout = setTimeout(() => controller.abort(), GUARD_TIMEOUT_MS);

      const res = await fetch(`${guardUrl}/v1/guard`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          Authorization: `Bearer ${guardKey}`,
        },
        body: JSON.stringify({ input: combined }),
        signal: controller.signal,
      });

      clearTimeout(timeout);

      if (res.ok) {
        const result = (await res.json()) as {
          classification: string;
          violationTypes?: string[];
        };

        if (result.classification === 'block') {
          console.warn('[Guard] Blocked injection in TokenForge request', {
            violations: result.violationTypes,
            tenantId: c.get('tenantId'),
            path: c.req.path,
          });

          return c.json({
            error: 'input_blocked',
            message: 'Request blocked by security guard',
            violations: result.violationTypes ?? [],
          }, 400);
        }
      }
    } catch {
      // Fail-open: guard unreachable, let request through
    }

    await next();
  };
}
